checkbox-ng/checkbox_ng/launcher/controller.py

```diff
@@ -354,7 +355,90 @@ def quitter(msg):
                 break
         return self._has_anything_failed
 
-    def resume_or_start_new_session(self):
+    def should_start_via_launcher(self):
+        """
+        Determines if the controller should automatically select a test plan
+        given a launcher. Raises if the launcher tries to skip the test plan
+        selection without providing the test plan that must be automatically
+        selected
+        """
+        if self.launcher.get_value("test plan", "forced"):
+            tp_unit = self.launcher.get_value("test plan", "unit")
+            if not tp_unit:
+                _logger.error(
+                    _(
+                        "The test plan selection was forced but no unit"
+                        " was provided"
+                    )
+                )
+                raise SystemExit(1)
+            return True
+        return False
+
+    @contextlib.contextmanager
+    def _resumed_session(self, session_id):
+        """
+        Used to temporarely resume a session to inspect it, abandoning it
+        before exiting the context
+        """
+        try:
+            yield self.sa.resume_session(session_id)
+        finally:
+            self.sa.abandon_session()
+
+    def should_start_via_autoresume(self) -> bool:
+        """
+        Determines if the controller should automatically resume a previously
+        abandoned session.
+
+        A session is automatically resumed if:
+        - A testplan was selected before abandoning
+        - A job was in progress when the session was abandoned
+        - The ongoing test was shell job
+        """
+        try:
+            last_abandoned_session = next(self.sa.get_resumable_sessions())
+        except StopIteration:
+            # no session to resume
+            return False
+        # resume session in agent to be able to peek at the latest job run
+        # info
+        # FIXME: IncompatibleJobError is raised if the resume candidate is
+        #        invalid, this is a workaround till get_resumable_sessions is
+        #        fixed
+        with contextlib.suppress(IncompatibleJobError), self._resumed_session(
+            last_abandoned_session.id
+        ) as metadata:
+            app_blob = json.loads(metadata.app_blob.decode("UTF-8"))
+
+            if not app_blob.get("testplan_id"):
+                self.sa.abandon_session()
+                return False
+
+            self.sa.select_test_plan(app_blob["testplan_id"])
+            self.sa.bootstrap()
+
+            if not metadata.running_job_name:
+                return False
+
+            job_state = self.sa.get_job_state(metadata.running_job_name)
+            if job_state.job.plugin != "shell":
+                return False
+            return True
```

Comment:
I'm wondering if it would be better to have this set of criteria could be outlined into a more testable and self contained function.

Reply:
I have written this in this format to avoid needless work on non-autoresumable sessions. Given how checkbox work we can not (to the best of my knowledge) inspect the session the way that we need without moving the state of the session via `select_tp`/`bootstrap`. Possibly the last part could be factored out given that I expect more constraints to be put in place on the job state potentially, but as long as it is only one, I would keep this here. I will go back on this either way to port this fix (potentially) to local